src/quickgui/qgsquickhighlightsgnode.cpp

```diff
@@ -0,0 +1,168 @@
+/***************************************************************************
+ qgsquickhighlightsgnode.cpp
+  --------------------------------------
+  Date                 : Nov 2017
+  Copyright            : (C) 2017 by Matthias Kuhn
+  Email                : matthias@opengis.ch
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsquickhighlightsgnode.h"
+#include "qgstessellator.h"
+#include "qgsmultipolygon.h"
+#include "qgsgeometrycollection.h"
+#include "qgsgeometry.h"
+#include "qgstriangle.h"
+
+QgsQuickHighlightSGNode::QgsQuickHighlightSGNode( const QgsGeometry &geom,
+    const QColor &color, float width )
+  : QSGNode()
+  , mWidth( width )
+{
+  mMaterial.setColor( color );
+  handleGeometryCollection( geom );
+}
+
+void QgsQuickHighlightSGNode::handleGeometryCollection( const QgsGeometry &geom )
+{
+  const QgsGeometryCollection *collection = qgsgeometry_cast<const QgsGeometryCollection *>( geom.constGet() );
+  if ( collection && !collection->isEmpty() )
+  {
+    for ( int i = 0; i < collection->numGeometries(); ++i )
+    {
+      QgsGeometry geomN( collection->geometryN( i )->clone() );
```

Comment:
This is a bit inefficient - if handleSingleGeometry took a `const QgsAbstractGeometry*` argument instead, you could avoid the clone